report/3.Theory/LexerImplementation.tex

```diff
@@ -0,0 +1,80 @@
+%This section covers the implementation of the lexer
+\section{Lexer implementation}
+As described in section \ref{LexerChoice}, it was decided that the lexer (and therefore the parser), would be written by hand.
+As mentioned in section \ref{REFTOPROGRAMLANGUAGECHOICE}, the compiler for HCL will be written in kotlin.
+Below is an explanation of the implementation of the lexer.
+
+The lexer is comprised of 3 different classes: the token class, the positional token class and the lexer class.
+
+The token class simply contains the type of the token. 
+It is implemented as a Sealed class. 
+A sealed class in Kotlin, simply specifies that a value can have one specific type, from a predefined set\cite{KotlinSealed}.
+It is essentially an extended enum.
+\begin{lstlisting}[language=java,label=lis:tokenClass,caption=A snippet from the token class .,firstnumber=9]
+sealed class Literal : Token() {
+	class Text(val value: String) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
+	}
+	class Number(strValue: String, val value: Double = strValue.toDouble()) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
+	}
+	class Bool(val value: Boolean) : Literal() {
+	override fun toString() = super.toString() + "[$value]"
+	}
+}
+\end{lstlisting}
+
+The positional token class is a wrapper class from a token, and simply contains additional information regarding linenumber and line index.
+The positional token is mainly implemented to help with error reporting.
+\begin{lstlisting}[language=java,label=lis:PositionalTokenClass,caption=A snippet from the token class .,firstnumber=10]
+class PositionalToken(val token: Token, val lineNumber: Int, val lineIndex: Int)
+\end{lstlisting}
+
+The bulk of the logic regarding the lexer is done in the lexer class.
+The lexer class implements the ILexer interface.
+
+\begin{lstlisting}[language=java,label=lis:Lexer,caption=The Lexer .,firstnumber=8]
+class Lexer(private val inputContent: String) : ILexer {
+\end{lstlisting}
+
+The ILexer defines a getTokenSequence function, which will give the token sequence and an inputLine function.

```

Comment:
lacks an oxford comma... right? or something - i read it as the getTokenSequence returned an input line function

Reply:
Yes, a comma after "sequence" would improve the sentence a lot